nautobot_netbox_importer/diffsync/nautobot.py

```diff
@@ -0,0 +1,362 @@
+            raise TypeError(f"Invalid diffsync type {self.diffsync}")
+        self.diffsync.save_nautobot_instance(instance, attrs, self._wrapper.fields)
+        return super().update(attrs)
+
+
+class NautobotAdapter(BaseAdapter):
+    """Nautobot DiffSync Adapter."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the adapter."""
+        super().__init__("Nautobot", *args, **kwargs)
+        self.clean_failures: Dict[NautobotBaseModelType, Set[Uid]] = {}
+        self.validation_errors: Optional[Dict[ContentTypeStr, Set[ValidationError]]] = None
+
+    def get_validation_errors(self) -> Dict[ContentTypeStr, Set[ValidationError]]:
+        """Re-run clean() on all instances that failed validation."""
+        if self.validation_errors is not None:
+            return self.validation_errors
+
+        self.validation_errors = {}
+
+        failures = self.clean_failures
+        self.clean_failures = {}
+
+        for model_type, uids in failures.items():
+            errors = None
+            for uid in uids:
+                instance = model_type.objects.get(id=uid)
+                try:
+                    instance.clean()
+                except ValidationError as error:
+                    if errors:
+                        errors.add(error)
+                    else:
+                        errors = set([error])
+                        self.validation_errors[model_type._meta.label.lower()] = errors  # type: ignore
+
+        return self.validation_errors
+
+    # pylint: disable=too-many-statements
+    def save_nautobot_instance(self, instance: NautobotBaseModel, values: RecordData, fields: NautobotFields) -> None:
+        """Save a Nautobot instance."""
+
+        def set_custom_field(value: Any):
+            custom_field_data = getattr(instance, "custom_field_data", None)
+            if custom_field_data is None:
+                raise TypeError("Missing custom_field_data")
+            custom_field_data.clear()
+            if value:
+                custom_field_data.update(value)
+
+        def set_generic_foreign_key(field: GenericForeignKey, value: Any):
+            if value:
+                foreign_model = get_model_from_name(value[0])
+                setattr(instance, field.ct_field, ContentType.objects.get_for_model(foreign_model))
+                setattr(instance, field.fk_field, value[1])
+            else:
+                setattr(instance, field.fk_field, None)
+                setattr(instance, field.ct_field, None)
+
+        def set_empty(field_name: FieldName, value: Any):
+            if value in EMPTY_VALUES:
+                if field.default not in EMPTY_VALUES:
+                    setattr(instance, field_name, field.default)
+                elif field.blank and not field.null:
+                    setattr(instance, field_name, "")
+                else:
+                    setattr(instance, field_name, None)
+            else:
+                setattr(instance, field_name, value)
```

Comment:
Seems weird for `set_empty` to have an else case for handling non-empty values?

Reply:
Fixed.